# 02 Data Preprocessing

Preprocessing CRLMSSold*.csv data according to findings from the EDA (01_exploration.ipynb) to create clean CSV and prep for week 4.

## Set Up

Loading the same .csv files used in 01_exploration.ipynb

In [1]:
import pandas as pd
import numpy as np
import glob

# %pip install geopandas
import geopandas as gpd
from shapely.geometry import Point

pd.set_option('display.max_columns', 100)

Data_dir = '../../data'
PATH = "../../data/DistrictAreas2526_-284845464123469011.geojson"

data = sorted(glob.glob(f'{Data_dir}/CRMLSSold*.csv'))

#error check
assert len(data) > 0, "No CRMLSSold*.csv files found."
print(f'Found {len(data)} file(s):')
for f in data:
    print(f'  {f}')

Found 13 file(s):
  ../../data\CRMLSSold202506.csv
  ../../data\CRMLSSold202507.csv
  ../../data\CRMLSSold202508.csv
  ../../data\CRMLSSold202509.csv
  ../../data\CRMLSSold202510.csv
  ../../data\CRMLSSold202511.csv
  ../../data\CRMLSSold202512.csv
  ../../data\CRMLSSold202601.csv
  ../../data\CRMLSSold202602.csv
  ../../data\CRMLSSold202603.csv
  ../../data\CRMLSSold202604.csv
  ../../data\CRMLSSold202605.csv
  ../../data\CRMLSSold202606.csv


In [2]:
all_csv = []
raw_total = 0

# Apply filters to all files
for file in data:
    #print(f"Processing file: {file}")
    csv = pd.read_csv(file, low_memory=False)
    raw_total += len(csv)

    # restrict analysis per task doc
    csv = csv[(csv['PropertyType'] == 'Residential') & (csv['PropertySubType'] == 'SingleFamilyResidence')].copy()
    
    all_csv.append(csv)

df = pd.concat(all_csv, ignore_index=True)

#geo check: drop out-of-scope rows (non-California)

CA_COUNTIES = {
    'Alameda','Alpine','Amador','Butte','Calaveras','Colusa','Contra Costa','Del Norte',
    'El Dorado','Fresno','Glenn','Humboldt','Imperial','Inyo','Kern','Kings','Lake','Lassen',
    'Los Angeles','Madera','Marin','Mariposa','Mendocino','Merced','Modoc','Mono','Monterey',
    'Napa','Nevada','Orange','Placer','Plumas','Riverside','Sacramento','San Benito',
    'San Bernardino','San Diego','San Francisco','San Joaquin','San Luis Obispo','San Mateo',
    'Santa Barbara','Santa Clara','Santa Cruz','Shasta','Sierra','Siskiyou','Solano','Sonoma',
    'Stanislaus','Sutter','Tehama','Trinity','Tulare','Tuolumne','Ventura','Yolo','Yuba',
}

n_before = len(df)
df = df[df['StateOrProvince'] == 'CA'].reset_index(drop=True)
df = df[df['CountyOrParish'].isin(CA_COUNTIES)].reset_index(drop=True)
print(f'Dropped {n_before - len(df):,} non-CA / out of scope')

# Reverify total files read
print(f'Total files read: {len(all_csv)}')
# Rows before filtering
print(f'Total rows read pre-filter: {raw_total:,}')
print(f'Total rows read post-filter (restrictions): {len(df):,}')

Dropped 12 non-CA / out of scope
Total files read: 13
Total rows read pre-filter: 283,176
Total rows read post-filter (restrictions): 143,066


### Convert CloseDate to datetime

Necessary step since deduplication and week 4 task requires validity and consistent real dates.

In [3]:
df['CloseDate'] = pd.to_datetime(df['CloseDate'], errors='coerce')

bad_dates = df['CloseDate'].isna().sum()
print(f'CloseDate parse failures: {bad_dates}')

# drop any date parse fails
if bad_dates > 0:
    df = df.dropna(subset=['CloseDate']).reset_index(drop=True)
    print(f'Dropped {bad_dates} rows with unparseable CloseDate.')

CloseDate parse failures: 0


In [4]:
#scalar timestamp

TEST_MONTH_START = df['CloseDate'].max().to_period('M').to_timestamp()
print(f'Test month starts: {TEST_MONTH_START.date()} '
      f'(rows on/after this are excluded from every learned statistic)')

Test month starts: 2026-06-01 (rows on/after this are excluded from every learned statistic)


## Deduplication

- EDA saw 99 duplicated ListingKeys
- Of the 99 duplicates, most pairs share the same ClosePrice, but conflict through CloseDates.
- Some pairs conflict on the ClosePrice

In [5]:
#quantify and inspect the conflict groups before touching anything

pre_dedup = len(df)
# print(f'amount of rows before deduplication: {pre_dedup}')

dup_rows = df[df['ListingKey'].duplicated(keep=False)]
price_nunique = dup_rows.groupby('ListingKey')['ClosePrice'].nunique()

conflict_keys = price_nunique[price_nunique > 1].index
print(f'Total Dupe ListingKeys: {len(price_nunique)}')
print(f'- Agreeing with ClosePrice: {(price_nunique == 1).sum()}')
print(f'- Conflict with ClosePrice: {len(conflict_keys)}')

Total Dupe ListingKeys: 99
- Agreeing with ClosePrice: 75
- Conflict with ClosePrice: 24


In [6]:
# Inspecting the conflicting groups
insp_cols = [col for col in ['ListingKey', 'ClosePrice', 'CloseDate', 'ListPrice']
                if col in df.columns]

display(df[df['ListingKey'].isin(conflict_keys)] .sort_values(['ListingKey', 'CloseDate'])[insp_cols])

,ListingKey,ClosePrice,CloseDate,ListPrice
11109,1100754427,364900.0,2025-06-22,364900.0
23474,1100754427,365000.0,2025-07-03,364900.0
68190,1110678316,290000.0,2025-11-07,285000.0
78702,1110678316,299000.0,2025-12-23,285000.0
32985,1112677801,289000.0,2025-08-25,289900.0
45354,1112677801,295000.0,2025-09-25,289900.0
57382,1114631996,842382.0,2025-10-10,766289.0
94763,1114631996,700000.0,2026-02-18,766289.0
29044,1118342365,700000.0,2025-08-22,719000.0
43169,1118342365,690882.0,2025-09-12,719000.0


### Dealing with Duplicates (ListingKey)

To deal with the duplicates that are re-reported in later month, we will use the recent version of the MLS as sorted using CloseDate (and ContractStatusChangeDate for verification).

In [7]:
sort_cols = ['ListingKey', 'CloseDate']
if 'ContractStatusChangeDate' in df.columns:
    sort_cols.append('ContractStatusChangeDate')

df = (df.sort_values(sort_cols)
        .drop_duplicates(subset='ListingKey', keep='last')
        .reset_index(drop=True))

print(f'Rows removed by dedup: {pre_dedup - len(df)}')
print(f'Remaining rows: {len(df):,}')
assert df['ListingKey'].is_unique

Rows removed by dedup: 100
Remaining rows: 142,966


### Implausible Values

Values that were flaged from the EDA

| Field | Problem | Count (pre-dedup) |
|---|---|---|
| ClosePrice | min = $1.75 reall small value | 7 |
| ClosePrice | max $796M extreme outlier | 18 |
| LivingArea | 0 size is uninhabitable| 53 |
| BedroomsTotal | value of 0, could be a studio needs clarifying | 77 |
| BathroomsTotalInteger | value of 0, not normal for residences| 61 |
| LotSizeAcres | 0 values, potential unit errors (43,560 = 1 acre in sq ft; max 60,113 acres) | 350 where LotSizeAcres = 0 |

In [8]:
# reverify EDA flags
checks = {
    'LivingArea == 0':            (df['LivingArea'] == 0).sum(),
    'BedroomsTotal == 0':         (df['BedroomsTotal'] == 0).sum(),
    'BathroomsTotalInteger == 0': (df['BathroomsTotalInteger'] == 0).sum(),
    'LotSizeAcres == 0':          (df['LotSizeAcres'] == 0).sum(),
    'ClosePrice < $10,000':       (df['ClosePrice'] < 10_000).sum(),
    'ClosePrice > $100M':         (df['ClosePrice'] > 100_000_000).sum(),
}
for label, n in checks.items():
    print(f'{label:<30} {n:>6,} rows')

LivingArea == 0                    53 rows
BedroomsTotal == 0                 77 rows
BathroomsTotalInteger == 0         61 rows
LotSizeAcres == 0                 350 rows
ClosePrice < $10,000                7 rows
ClosePrice > $100M                 18 rows


 ### i. Zero-value structure fields

 From above, it's difficult to accept that a single-family residence is missing a bedroom, bathroom, and is uninhabitable (LivingArea). These are more likely to be data entry issues rather than true values.

Decicion: Drop zero-value entries of BedroomsTotal, BathroomIntegersTotal, and LivingArea.

In [9]:
n_before = len(df)
print(n_before)

implausible_structure = (
    (df['LivingArea'] == 0) |
    (df['BedroomsTotal'] == 0) |
    (df['BathroomsTotalInteger'] == 0)
)
df = df[~implausible_structure].reset_index(drop=True)

print(f'Rows dropped for zero-value fields: {n_before - len(df):,}')
print(f'Remaining rows: {len(df):,}')

142966
Rows dropped for zero-value fields: 125
Remaining rows: 142,841


### ii. ClosePrice: Establishing boundaries

Some of the ClosePrice outliers can only be removed due to the impact on the rest of the distribution. The following code investigates the flagged entries each potential issue causing the flags:
- unit calculation errors
- abiguity

In [10]:
# Inspect the outlier rows
MIN = 10_000
MAX = 100_000_000

#Plausibilty/further cleaning
PPSF_MAX = 15_000


out_of_bounds = ~df['ClosePrice'].between(MIN, MAX)
implausible_ppsf = (df['ClosePrice'] / df['LivingArea']) > PPSF_MAX #revision
flagged = out_of_bounds | implausible_ppsf #revision

inspect_cols = [col for col in ['ListingKey', 'ClosePrice', 'ListPrice',
                                'LivingArea', 'CountyOrParish', 'City', 'CloseDate']
                if col in df.columns]
inspect = df.loc[flagged, inspect_cols].copy()
inspect['Close/List'] = (inspect['ClosePrice'] / inspect['ListPrice']).round(4)
inspect['$/sqft'] = (inspect['ClosePrice'] / inspect['LivingArea']).round(0) #revision

print(f'Rows outside [${MIN:,}, ${MAX:,}]: {out_of_bounds.sum()}')
display(inspect.sort_values('ClosePrice'))

Rows outside [$10,000, $100,000,000]: 25


,ListingKey,ClosePrice,ListPrice,LivingArea,CountyOrParish,City,CloseDate,Close/List,$/sqft
33562,1117019091,0.000000e+00,1350000.0,2001.0,Los Angeles,Temple City,2025-07-30,0.0000,0.0
81483,1144223437,1.750000e+00,1749000.0,3513.0,Riverside,Palm Desert,2026-01-30,0.0000,0.0
61899,1129501416,4.850000e+02,498000.0,1305.0,Riverside,Cathedral City,2025-10-23,0.0010,0.0
79652,1140167514,5.800000e+02,599000.0,1910.0,Riverside,Palm Desert,2026-06-18,0.0010,0.0
104022,1151459976,6.850000e+02,913655.0,2980.0,Contra Costa,Oakley,2026-04-29,0.0007,0.0
93297,1147508351,8.000000e+03,8000.0,984.0,San Bernardino,Trona,2026-02-04,1.0000,8.0
8925,1110035161,8.300000e+03,2000000.0,3886.0,Los Angeles,Altadena,2026-01-17,0.0042,2.0
2581,1103309672,2.200000e+07,45000000.0,1200.0,Orange,Laguna Beach,2025-07-11,0.4889,18333.0
63704,1131041267,2.370000e+07,2495000.0,1428.0,Santa Barbara,Santa Barbara,2025-10-15,9.4990,16597.0
120189,1154925192,4.872000e+07,5150000.0,1800.0,San Diego,Encinitas,2026-05-15,9.4602,27067.0


In [11]:
# Potential unit corrections

FACTORS = [1, 1e-3, 1e-2, 1e-1, 1e3, 1e6]

# 1 = keep, 
# 1e-3 = divide by 1,000 

#revision: added more factors to check for potential unit errors
# 1e-2 = divide by 100
# 1e-1 = divide by 10

# 1e3 = multiply by 1,000
# 1e6 = multiply by 1M

RATIO = (0.80, 1.25) 

out= df.index[flagged]
df['ClosePrice_repaired'] = 0

repaired, kept, dropped = [], [], []

for i in out:
    price = df.at[i, 'ClosePrice']
    lst   = df.at[i, 'ListPrice'] #compare to the listing price to see if the price is a unit error
    if pd.isna(lst) or lst <= 0: #doesn't have a valid comparing price so drop
        dropped.append((i, price, lst, []))
        continue

    matches = []
    for f in FACTORS:
        corrected = price * f
        ratio_ok  = RATIO[0] <= corrected / lst <= RATIO[1]
        bounds_ok = (f == 1) or (MIN <= corrected <= MAX)
        if ratio_ok and bounds_ok:
            matches.append(f)

    if matches == [1]:
        kept.append((i, price, lst)) #price is already in bounds and ratio is okay
    elif len(matches) == 1:
        df.at[i, 'ClosePrice'] = price * matches[0]
        df.at[i, 'ClosePrice_repaired'] = 1
        repaired.append((i, price, price * matches[0], matches[0]))
    else: # IS ABIGUOUS CASE
        dropped.append((i, price, lst, matches))
    
print(f'Out-of-bounds rows examined: {len(out)}\n')
print(f'# REPAIRED rows (via unit decoding): {len(repaired)}\n')

for i, old, new, factor in repaired:
    key = df.at[i, 'ListingKey']
    lst = df.at[i, 'ListPrice']
    closed = df.at[i, 'CloseDate'].date()
    ratio = new / lst
    status = "PASS" if RATIO[0] <= ratio <= RATIO[1] else "FAIL"

    print(f' ListingKey {key} | CloseDate: {closed}:\n'
          f'    ListPrice: ${lst:,.0f} \n'
          f'    Original ClosePrice: ${old:,.2f} \n'
          f'    --> Calculation: ${old:,.2f} × {factor:g} \n'
          f'    ==> Repaired ClosePrice: ${new:,.0f} \n' 
          f'    Ratio Check: {new:,.0f} / {lst:,.0f} = {ratio:.3f} ({status})\n')
    
print(f'# REMAINING rows as Original ClosePrice: {len(kept)}')

for i, price, lst in kept:
    key = df.at[i, 'ListingKey']
    closed = df.at[i, 'CloseDate'].date()
    ratio = price / lst
    status = "PASS" if RATIO[0] <= ratio <= RATIO[1] else "FAIL"

    print(f'ListingKey {key} | CloseDate: {closed}:\n'
          f'    ListPrice: ${lst:,.0f}\n'
          f'    Original ClosePrice (Kept): ${price:,.2f}\n'
          f'    Ratio Check: {price:,.0f} / {lst:,.0f} = {ratio:.3f} ({status})\n')


print(f'# DROPPED rows (ambiguous): {len(dropped)}')

for i, price, lst, matches in dropped:
    key = df.at[i, 'ListingKey']
    closed = df.at[i, 'CloseDate'].date()

    print(f'ListingKey {key} | CloseDate: {closed}:\n'
        f'    ListPrice: ${lst:,.0f}\n'
        f'    Original ClosePrice: ${price:,.2f}\n'
        f'    Matching Factors: {matches if matches else "None"}\n'
        f'    Decision: DROPPED (ambiguous or no valid correction)\n')

drop_i = [i for i, _, _, _ in dropped]
n_before = len(df)
df = df.drop(index=drop_i).reset_index(drop=True)
print(f'Rows dropped: {n_before - len(df)} | Total Rows Remaining in df: {len(df):,}')

Out-of-bounds rows examined: 29

# REPAIRED rows (via unit decoding): 21

 ListingKey 1115065523 | CloseDate: 2025-09-10:
    ListPrice: $809,000 
    Original ClosePrice: $815,000,000.00 
    --> Calculation: $815,000,000.00 × 0.001 
    ==> Repaired ClosePrice: $815,000 
    Ratio Check: 815,000 / 809,000 = 1.007 (PASS)

 ListingKey 1116006224 | CloseDate: 2025-07-28:
    ListPrice: $649,000 
    Original ClosePrice: $613,000,000.00 
    --> Calculation: $613,000,000.00 × 0.001 
    ==> Repaired ClosePrice: $613,000 
    Ratio Check: 613,000 / 649,000 = 0.945 (PASS)

 ListingKey 1125862013 | CloseDate: 2025-09-11:
    ListPrice: $835,800 
    Original ClosePrice: $835,800,000.00 
    --> Calculation: $835,800,000.00 × 0.001 
    ==> Repaired ClosePrice: $835,800 
    Ratio Check: 835,800 / 835,800 = 1.000 (PASS)

 ListingKey 1126046881 | CloseDate: 2025-11-24:
    ListPrice: $735,000 
    Original ClosePrice: $740,000,000.00 
    --> Calculation: $740,000,000.00 × 0.001 
    ==> Repa

Notes:

- 21/29 rows repaired due to unit inconsistencies
- 2/29 rows maintained as-is due to reasonable connection to its ListPrice.
- 6/29 rows dropped entirely due to ambiguity

### iii. LotSizeAcres: 

The following strategy investigates the flagged entries each potential issue causing the flags:
- extreme skew
- log transform does not fix
- unit-conversion errors/consistency

Using LotSizeSquare to offer reference and potential solutions.

In [12]:
# Cross check units through LotSizeAcres vs LotSizeSquare calculations

if 'LotSizeSquareFeet' not in df.columns:
    print('LotSizeSquareFeet missing, skipping cross-check.')
else:
    acres = df['LotSizeAcres']
    sqft  = df['LotSizeSquareFeet']

    acres_ok = acres.notna() & (acres > 0)
    sqft_ok  = sqft.notna() & (sqft > 0)
    both     = acres_ok & sqft_ok

    consistent     = both & np.isclose(sqft, acres * 43_560, rtol=0.01)
    copy_error     = both & ~consistent & np.isclose(acres, sqft, rtol=0.001)
    mismatch = both & ~consistent & ~copy_error
    repair_src  = ~acres_ok & sqft_ok #acres unusable, sqft available
    both_unusable  = ~acres_ok & ~sqft_ok

    print(f'Rows both populated & nonzero:     {both.sum():>7,}')
    print(f'Consistent (ratio ~43,560):        {consistent.sum():>7,}')
    print(f'Copy error (acres == sqft):        {copy_error.sum():>7,}')
    print(f'Other mismatch:                    {mismatch.sum():>7,}')
    print(f'\nLotSizeAcres unusable, LotSizeSquareFeet available: {repair_src.sum():>7,}')
    print(f'Both variables unusable:           {both_unusable.sum():>7,}')

    # Flags as suspected sq-ft entries:
    suspected = acres > 640
    print(f'\nSuspected sq-ft entries (acres > 640):   {suspected.sum():,}')
    print(f'  - copy errors (sqft == acres):         {(suspected & copy_error).sum():,}')
    print(f'  = "consistent" :                       {(suspected & consistent).sum():,}')

    if mismatch.sum() > 0:
        print('\nSample of mismatches:')
        display(df.loc[mismatch,
                       ['ListingKey', 'LotSizeAcres', 'LotSizeSquareFeet']].head(10))

Rows both populated & nonzero:     140,029
Consistent (ratio ~43,560):        140,017
Copy error (acres == sqft):              0
Other mismatch:                         12

LotSizeAcres unusable, LotSizeSquareFeet available:     275
Both variables unusable:             2,521

Suspected sq-ft entries (acres > 640):   115
  - copy errors (sqft == acres):         0
  = "consistent" :                       105

Sample of mismatches:


,ListingKey,LotSizeAcres,LotSizeSquareFeet
3294,1106380600,0.0001,4.00
5181,1108306662,0.0001,5.23
22948,1114175913,0.0002,8.46
38072,1118315264,0.0001,3.07
45433,1119094165,0.0006,25.00
49109,1119732788,0.0003,13.80
52416,1120172649,0.0002,6.72
60772,1128444387,0.0026,112.00
83360,1144778109,0.0003,11.76
89979,1146526915,0.0001,2.20


In [13]:
MIN_ACRES = 0.01   # ~436 sq ft
MAX_ACRES = 640    # 1 sq-mi
SQFT_PER_ACRE = 43_560

lot = df['LotSizeAcres'].copy()
og_na = lot.isna().sum()

#zero lot size -> NaN
zero = lot.index[lot == 0]
lot.loc[zero] = np.nan
print(f'Zero (0) lot size -> NaN: {len(zero)} rows')

#suspected sq-ft entries
suspect_idx = lot.index[lot > MAX_ACRES]
print(f'\nSuspected sq-ft entries (> {MAX_ACRES} acres): {len(suspect_idx)} rows\n')

repaired = 0
for i in suspect_idx:
    val  = lot.at[i]
    conv = val / SQFT_PER_ACRE
    ok   = MIN_ACRES <= conv <= MAX_ACRES
    key    = df.at[i, 'ListingKey']
    closed = df.at[i, 'CloseDate'].date()

    # Cross-reference sq-ft colum:
    # sqft == acres value
    # sqft == acres x 43,560

    xref = ''
    if 'LotSizeSquareFeet' in df.columns:
        sqft = df.at[i, 'LotSizeSquareFeet']
        if pd.isna(sqft):
            xref = '    Cross-ref LotSizeSquareFeet: missing\n'
        elif np.isclose(sqft, val, rtol=0.001):
            xref = (f'    Cross-ref LotSizeSquareFeet: {sqft:,.0f} '
                    f'== acres value (COPY ERROR confirmed)\n')
        elif np.isclose(sqft, val * SQFT_PER_ACRE, rtol=0.01):
            xref = (f'    Cross-ref LotSizeSquareFeet: {sqft:,.0f} '
                    f'= acres x 43,560 (auto-derived; sq-ft column equally corrupted)\n')
        else:
            xref = (f'    Cross-ref LotSizeSquareFeet: {sqft:,.0f} '
                    f'(inconsistent with both readings)\n')

    status   = 'PASS' if ok else 'FAIL'
    decision = f'REPAIRED -> {conv:.4f} acres' if ok else 'set to NaN (conversion implausible)'
    print(f'  ListingKey {key} | CloseDate: {closed}:\n'
          f'    Original LotSizeAcres: {val:,.4f}\n'
          + xref +
          f'    --> Calculation: {val:,.4f} / {SQFT_PER_ACRE:,}\n'
          f'    ==> Converted: {conv:.4f} acres\n'
          f'    Plausibility Check: {MIN_ACRES} <= {conv:.4f} <= {MAX_ACRES} ({status})\n'
          f'    Decision: {decision}\n')

    if ok:
        lot.at[i] = conv
        repaired += 1
    else:
        lot.at[i] = np.nan

# small nonzero values -> NaN
small = lot.index[lot < MIN_ACRES]
print(f'Small (> 0, < {MIN_ACRES} acres) -> NaN: {len(small)} rows')
SHOW = 10

for i in small[:SHOW]:
    key = df.at[i, 'ListingKey']
    print(f'  ListingKey {key}: {lot.at[i]:.6f} acres (~{lot.at[i]*SQFT_PER_ACRE:,.0f} sq ft) -> NaN')
if len(small) > SHOW:
    print(f'  ... and {len(small) - SHOW} more')
lot.loc[small] = np.nan

# recover from LotSizeSquareFeet where acres is unusable
recovered, rejected = 0, 0
if 'LotSizeSquareFeet' in df.columns:
    cand = lot.index[lot.isna()
                     & df['LotSizeSquareFeet'].notna()
                     & (df['LotSizeSquareFeet'] > 0)]
    print(f'\nAcres unusable, sqft populated: {len(cand)} candidate rows')
    shown = 0
    for i in cand:
        sqft = df.at[i, 'LotSizeSquareFeet']
        conv = sqft / SQFT_PER_ACRE
        ok   = MIN_ACRES <= conv <= MAX_ACRES
        if shown < SHOW:
            key = df.at[i, 'ListingKey']
            verdict = f'RECOVERED: {conv:.4f} acres' if ok else 'rejected (outside plausible range)'
            print(f'  ListingKey {key}: {sqft:,.0f} sq ft / {SQFT_PER_ACRE:,} = {conv:.4f} -> {verdict}')
            shown += 1
        if ok:
            lot.at[i] = conv
            recovered += 1
        else:
            rejected += 1
    if len(cand) > SHOW:
        print(f'and {len(cand) - SHOW} more candidates')
    print(f' Recovered: {recovered} | Rejected: {rejected}')

df['LotSizeAcres'] = lot
df['LotSizeAcres_imputed'] = lot.isna() #flag

Zero (0) lot size -> NaN: 350 rows

Suspected sq-ft entries (> 640 acres): 115 rows

  ListingKey 1014091927 | CloseDate: 2026-06-26:
    Original LotSizeAcres: 2,452.6300
    Cross-ref LotSizeSquareFeet: 106,836,563 = acres x 43,560 (auto-derived; sq-ft column equally corrupted)
    --> Calculation: 2,452.6300 / 43,560
    ==> Converted: 0.0563 acres
    Plausibility Check: 0.01 <= 0.0563 <= 640 (PASS)
    Decision: REPAIRED -> 0.0563 acres

  ListingKey 1034539730 | CloseDate: 2025-10-15:
    Original LotSizeAcres: 1,402.5200
    Cross-ref LotSizeSquareFeet: 61,093,771 = acres x 43,560 (auto-derived; sq-ft column equally corrupted)
    --> Calculation: 1,402.5200 / 43,560
    ==> Converted: 0.0322 acres
    Plausibility Check: 0.01 <= 0.0322 <= 640 (PASS)
    Decision: REPAIRED -> 0.0322 acres

  ListingKey 1075488875 | CloseDate: 2025-06-20:
    Original LotSizeAcres: 1,605.0000
    Cross-ref LotSizeSquareFeet: 69,913,800 = acres x 43,560 (auto-derived; sq-ft column equally corrupte

In [14]:
# Summary
print(f'\nSummary:')
print(f'  Originally missing:            {og_na:,}')
print(f'  Zeros -> NaN:                  {len(zero):,}')
print(f'  Sq-ft entries repaired:        {repaired:,}')
print(f'  Suspected but unconvertible:   {len(suspect_idx) - repaired:,}')
print(f'  Too-small -> NaN:              {len(small):,}')
print(f'  Recovered from sqft column:    {recovered:,}')
print(f'  Total NaN awaiting imputation: {lot.isna().sum():,} ({lot.isna().mean()*100:.2f}% of rows)')
print(f'  Post-repair max LotSizeAcres:  {lot.max():,.2f}')


Summary:
  Originally missing:            2,446
  Zeros -> NaN:                  350
  Sq-ft entries repaired:        115
  Suspected but unconvertible:   0
  Too-small -> NaN:              26
  Recovered from sqft column:    0
  Total NaN awaiting imputation: 2,822 (1.98% of rows)
  Post-repair max LotSizeAcres:  552.58


Notes: 
- The LotSizeSquareFeet variable served as a cross-check, not a repair source
    - it is auto-derived from acres, so the 115 unit-error entries were repaired via the ÷43,560 heuristic (all 115 successfully, validated by their conversions landing in the normal 0.05–1.38 acre range).
- 350 zero-valued entries and 26 implausibly small lot sizes were recoded as missing (NaN) since could not be reliably interpreted.
- No suspected square-foot entries remained unconvertible; all 45 identified unit inconsistencies were successfully repaired.
- attempted recovery of unusable lot sizes from LotSizeSquareFeet (275 candidates) recovered 0 every candidate value was near-zero junk rejected by the plausibility guard, confirming the column carries no independent information.
- The 2,822 remaining NaN entries are flagged and passed to imputation.

## Missing Values

Handling missing values as found in 01_exploration.ipynb
- decide whether to drop, impute, or flag

### i. DROP

Dropping columns that are 100% null dynamically for any dataset.
Re-performance of the missingness column of 01_exploration.ipynb.

In [15]:
cols = df.columns[df.isna().all()].tolist()

print(f'Dropping {len(cols)} columns that are 100% null:')
for col in cols:
    print(f'  {col}')

df = df.drop(columns=cols)
print(f'\nColumns remaining: {df.shape[1]}')

Dropping 8 columns that are 100% null:
  FireplacesTotal
  AboveGradeFinishedArea
  TaxAnnualAmount
  TaxYear
  ElementarySchoolDistrict
  BusinessType
  CoveredSpaces
  MiddleOrJuniorSchoolDistrict

Columns remaining: 72


Note:
- Dynamically identified and dropped 8 columns that contained 100% missing values
- The pipeline automatically adapts to future datasets with different fully-null columns.
- 72 columns remained after removing the fully-null features.

### ii. Boolean Fields

- standardizing NaN by setting as 'False' for bool columns


In [16]:
#check for bool cols
yn_cols = [col for col in df.columns if col.endswith('YN')]
print(f'{len(yn_cols)} YN columns found: {yn_cols}\n')

# standardizing the representations
truthy_map = {True: True, False: False, 'True': True, 'False': False,
              'Y': True, 'N': False, 1: True, 0: False, 1.0: True, 0.0: False}
for col in yn_cols:
    df[col] = df[col].map(truthy_map)
    df[col] = df[col].fillna(False).astype(bool)

MINORITY = 0.001  # minority class

drop_yn = []
for col in yn_cols:
    true_share = df[col].mean()
    minority = min(true_share, 1 - true_share)
    status = 'DROP (near-constant)' if minority < MINORITY else 'keep'
    print(f'  {col:<28} True: {true_share*100:6.2f}%   -> {status}')
    if minority < MINORITY:
        drop_yn.append(col)

df = df.drop(columns=drop_yn)
print(f'\nDropped {len(drop_yn)} near-constant YN columns | Columns remaining: {df.shape[1]}')

7 YN columns found: ['ViewYN', 'WaterfrontYN', 'BasementYN', 'PoolPrivateYN', 'AttachedGarageYN', 'FireplaceYN', 'NewConstructionYN']

  ViewYN                       True:  56.25%   -> keep
  WaterfrontYN                 True:   0.05%   -> DROP (near-constant)
  BasementYN                   True:   2.42%   -> keep
  PoolPrivateYN                True:  15.57%   -> keep
  AttachedGarageYN             True:  73.50%   -> keep
  FireplaceYN                  True:  72.86%   -> keep
  NewConstructionYN            True:   3.40%   -> keep

Dropped 1 near-constant YN columns | Columns remaining: 71


Note:

- Standardized 7 YN indicator columns into consistent Boolean (True/False) values by mapping mixed representations (Y/N, True/False, 1/0) to a common format.
- Dropped 1 near-constant Boolean feature (WaterfrontYN) after identifying a minority class frequency below the 0.1% threshold.
- Retained the remaining 6 Boolean features, as each contained sufficient variation to provide potential predictive value.
- 71 columns remained after removing the near-constant feature.

### iii. Missing strong predictors

- LivingArea / BedroomsTotal / BathroomsTotalInteger
- checking for missingness again and dropping from the dataset

In [17]:
n_before = len(df)
df = df.dropna(subset=['LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger']).reset_index(drop=True)
print(f'Rows dropped for missing key predictors: {n_before - len(df):,}')
print(f'Remaining rows: {len(df):,}')

Rows dropped for missing key predictors: 85
Remaining rows: 142,750


- These features are considered essential predictors for modeling housing prices, so incomplete records were removed rather than imputed.
- 142,750 rows remained after filtering for complete key predictor information.

### iv. LotSizeAcres: Impute test


In [18]:
# First day of the most recent month in the data = start of the (future) test period
# test_month_start = df['CloseDate'].max().to_period('M').to_timestamp()
# fit_mask = df['CloseDate'] < test_month_start
fit_mask = df['CloseDate'] < TEST_MONTH_START

lot_median = df.loc[fit_mask, 'LotSizeAcres'].median()
n_filled = df['LotSizeAcres'].isna().sum()

df['LotSizeAcres'] = df['LotSizeAcres'].fillna(lot_median)

print(f'Test month starts: {TEST_MONTH_START.date()} (rows on/after this date excluded from the median)')
print(f'Median LotSizeAcres (fit on pre-test-month rows): {lot_median:.4f} acres')
print(f'Rows imputed: {n_filled:,} (flagged in LotSizeAcres_imputed)')

Test month starts: 2026-06-01 (rows on/after this date excluded from the median)
Median LotSizeAcres (fit on pre-test-month rows): 0.1674 acres
Rows imputed: 2,821 (flagged in LotSizeAcres_imputed)


Note:
- Imputed missing LotSizeAcres values using the median calculated from pre-test-month observations
- All imputed values were flagged in the LotSizeAcres_imputed indicator column to preserve transparency for downstream analysis.
- counts are scaled with the months in the df so can change depending on what's given

### v. Missingness

This audit lists what still contains NaNs, seeding directly into the keep/drop
decisions of the next section.

In [19]:
remaining = (df.isna().mean() * 100).round(2)
remaining = remaining[remaining > 0].sort_values(ascending=False)

if len(remaining) == 0:
    print('No missing values.')
else:
    print(f'{len(remaining)} columns still contain NaNs (% missing):')
    print(remaining.to_string())

42 columns still contain NaNs (% missing):
BelowGradeFinishedArea     99.29
BuilderName                95.47
LotSizeDimensions          93.70
BuildingAreaTotal          93.34
CoBuyerAgentFirstName      90.58
ElementarySchool           86.97
MiddleOrJuniorSchool       86.84
HighSchool                 82.84
CoListAgentFirstName       76.62
CoListAgentLastName        76.54
CoListOfficeName           75.10
AssociationFeeFrequency    74.43
SubdivisionName            64.53
MainLevelBedrooms          39.18
Flooring                   35.51
AssociationFee             29.10
HighSchoolDistrict         26.96
MLSAreaMajor               14.69
Stories                    10.44
BuyerOfficeAOR              8.37
Levels                      7.25
GarageSpaces                3.88
BuyerAgentAOR               3.52
LotSizeSquareFeet           1.71
LotSizeArea                 1.70
BuyerOfficeName             1.14
ListAgentFirstName          0.67
BuyerAgentFirstName         0.43
ListAgentEmail              0.27


Note:

- 42 columns still contained some missing values, with the highest missingness concentrated in optional/descriptive fields such as BelowGradeFinishedArea, BuilderName, LotSizeDimensions, and school-related columns.
- This audit is used to guide the next keep/drop decisions rather than immediately removing all columns with missing values.

## 4. Feature Selection & Leakage

- Dealing with data leakage prevention
- unique values
- Non or weak correlations with the target variable


### i. Preventing Data Leakage

In [20]:
price_like = [col for col in df.columns
              if 'Price' in col and col not in ('ClosePrice', 'ClosePrice_repaired')]
print(f'Dropping {len(price_like)} price-like columns as leakage:')
for col in price_like:
    print(f'  {col}')
df = df.drop(columns=price_like)

Dropping 2 price-like columns as leakage:
  OriginalListPrice
  ListPrice


### ii. Explicit keep


In [21]:
id_cols      = ['ListingKey', 'CloseDate']
target_cols  = ['ClosePrice']
feature_cols = ['LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger','LotSizeAcres', 'LotSizeAcres_imputed', 'YearBuilt', 'Stories', 'GarageSpaces', 'ParkingTotal'] #REVISION: expansion
yn_cols      = [col for col in df.columns if col.endswith('YN')]  #sourced from 3ii.
cat_cols     = ['CountyOrParish', 'City']
ref_cols     = ['PostalCode', #raw ref
                'ClosePrice_repaired']
geo_cols     = ['Latitude', 'Longitude']   # REVISION

keep = id_cols + target_cols + feature_cols + yn_cols + cat_cols + ref_cols + geo_cols  # REVISION: added geo_cols
not_found = [col for col in keep if col not in df.columns]
assert not not_found, f'Expected columns missing from this extract: {not_found}'

dropped = [col for col in df.columns if col not in keep]
print(f'Dropping remaining non-feature columns: {len(dropped)}')

df = df[keep].copy()
print(f'Columns kept ({df.shape[1]}):')
for col in df.columns:
    print(f'  {col}')

Dropping remaining non-feature columns: 45
Columns kept (24):
  ListingKey
  CloseDate
  ClosePrice
  LivingArea
  BedroomsTotal
  BathroomsTotalInteger
  LotSizeAcres
  LotSizeAcres_imputed
  YearBuilt
  Stories
  GarageSpaces
  ParkingTotal
  ViewYN
  BasementYN
  PoolPrivateYN
  AttachedGarageYN
  FireplaceYN
  NewConstructionYN
  CountyOrParish
  City
  PostalCode
  ClosePrice_repaired
  Latitude
  Longitude


In [22]:
CA_LAT, CA_LON = (32.3, 42.1), (-124.5, -114.0)      # California bounding box
bad = ~(df['Latitude'].between(*CA_LAT) & df['Longitude'].between(*CA_LON))
df.loc[bad, ['Latitude', 'Longitude']] = np.nan

##Revision: medians should be coming from the training period
fit_mask = df['CloseDate'] < TEST_MONTH_START

for coord in ['Latitude', 'Longitude']:
    county_medians = df.loc[fit_mask].groupby('CountyOrParish')[coord].median()
    global_median  = df.loc[fit_mask, coord].median()
    df[coord] = df[coord].fillna(df['CountyOrParish'].map(county_medians))
    df[coord] = df[coord].fillna(global_median)  # fallback if a county is all-NaN

assert df[['Latitude', 'Longitude']].isna().sum().sum() == 0, 'Coordinate NaNs remain after imputation'
print(f'coords: {int(bad.sum())} out-of-CA/missing rows imputed by county median')

coords: 52 out-of-CA/missing rows imputed by county median


In [23]:
#REVISION: whole block adding in the expanded features screening early on in model
# pretest = df['CloseDate'] < df['CloseDate'].max().to_period('M').to_timestamp()
pretest = df['CloseDate'] < TEST_MONTH_START

before = df[['YearBuilt', 'Stories', 'GarageSpaces', 'ParkingTotal']].isna().sum()

# YearBuilt screening
yr_max = int(df['CloseDate'].dt.year.max())
bad_year = df['YearBuilt'].isna() | ~df['YearBuilt'].between(1850, yr_max)
df['YearBuilt_imputed'] = bad_year
df.loc[bad_year, 'YearBuilt'] = np.nan
df['YearBuilt'] = df['YearBuilt'].fillna(df.loc[pretest, 'YearBuilt'].median())

# Stories: pretest median
df['Stories'] = df['Stories'].fillna(df.loc[pretest, 'Stories'].median())

# GarageSpaces / ParkingTotal: missing == 0 (recognize as no garage/parking ammenity)
for c in ['GarageSpaces', 'ParkingTotal']:
    df[c] = df[c].fillna(0)

_new = ['YearBuilt', 'Stories', 'GarageSpaces', 'ParkingTotal']
assert df[_new].isna().sum().sum() == 0, 'Structural NaNs remain after imputation'
print('missing before impute:')
print(before.to_string())
print(f"YearBuilt implausible/missing flagged: {int(bad_year.sum()):,}")

missing before impute:
YearBuilt          72
Stories         14899
GarageSpaces     5538
ParkingTotal        1
YearBuilt implausible/missing flagged: 77


### iii. PostalCode: ensuring consistency

In [24]:
df['PostalCode'] = (df['PostalCode'].astype('string')
                    .str.replace(r'\.0$', '', regex=True))
print(df['PostalCode'].head())

0    92103
1    92647
2    96094
3    90003
4    92352
Name: PostalCode, dtype: string


Notes:

- i. Preventing Data Leakage
    - Dropped 2 price-related variables (OriginalListPrice and ListPrice) to prevent data leakage during model training.
    - Retained ClosePrice as the target variable and ClosePrice_repaired as an audit flag indicating whether the target value was unit-corrected.
- ii. Explicit Keep
    - Explicitly retained 24 variables required for modeling, including the target, selected predictors, Boolean indicators, categorical features, and audit/reference columns.
    - Dropped the remaining 45 non-feature columns to create a streamlined modeling dataset.
    - PostalCode and ClosePrice_repaired were retained for future feature engineering and auditing but are not included as predictive features in the current baseline model.
- iii. PostalCode
    - Standardized and converted PostalCode to str data type to preserve leading zeros and ensure consistent formatting, also removing trailing decimal notations


# Expanded Feature Engineering

Week 6 Task:
- bed/bath ratio
- age of property in years
- detailed geographic layer using school districts

## School District Mapping
- downloaded 2025-26 GeoJSON, stored in /data (.gitignored) not uploaded to github
    - .../data/DistrictAreas2526_-284845464123469011.geojson
- Spatial join of each property's coordinates against the 2025-26 CA unified school district boundaries
- Unmatched rows (areas served by separate elementary + high districts, or edge/offshore points) fall back to a county-level label so they keep geographic resolution instead of collapsing into one bucket.

In [25]:
# Inpecting the .geojson diagnostics

# load
districts = gpd.read_file(PATH)
candidate = [c for c in districts.columns if "type" in c.lower() or "name" in c.lower() or "district" in c.lower()]
print("\ncandidate type/name columns:", candidate)
for c in candidate:
    print(districts[c].value_counts().head(8))


candidate type/name columns: ['CountyName', 'DistrictName', 'DistrictType']
CountyName
Los Angeles       79
Kern              46
Tulare            43
San Diego         42
Sonoma            40
San Bernardino    33
Fresno            31
Santa Clara       31
Name: count, dtype: int64
DistrictName
Jefferson Elementary        3
Pioneer Union Elementary    2
Pacific Union Elementary    2
Washington Unified          2
Lakeside Union              2
Mountain View Elementary    2
Ocean View                  2
Hope Elementary             2
Name: count, dtype: int64
DistrictType
Elementary    515
Unified       345
High           76
Name: count, dtype: int64


In [26]:
# Filter to unified districts
unified = districts[districts["DistrictType"] == "Unified"].copy()
print(f"unified districts: {len(unified)}")

# Convert each property’s Latitude and Longitude into a geographic point
pts = gpd.GeoDataFrame(
    df.copy(),
    geometry=[Point(xy) for xy in zip(df["Longitude"], df["Latitude"])],
    crs="EPSG:4326",
).to_crs(unified.crs)

# spatial join
keep_cols = ["DistrictName", "CountyName", "geometry"]
joined = gpd.sjoin(pts, unified[keep_cols], how="left", predicate="within")

# check for duplicates
dupes = joined.index.duplicated().sum()
if dupes:
    print(f"WARNING: {dupes} points matched multiple polygons; keeping first")
    joined = joined[~joined.index.duplicated(keep="first")]

unified districts: 345


In [27]:
# dealing with shared names
name_counts = unified["DistrictName"].value_counts()
shared = set(name_counts[name_counts > 1].index)

dist = joined["DistrictName"]
d_cty = joined["CountyName"] #given districtcounty in geojson
fbcty  = df["CountyOrParish"] #fallback of its own county

# matched + shared name -> "Name (County)"; matched + unique -> "Name"
label = dist.where(~dist.isin(shared), dist + " (" + d_cty.fillna("") + ")")

# unmatched
label = label.where(dist.notna(), "NonUnified_" + fbcty.astype(str))
df["SchoolDistrict"] = label.values

# check
matched = dist.notna()

assert df["SchoolDistrict"].isna().sum() == 0, "SchoolDistrict has NaNs"
print(f"matched:   {matched.sum():,} of {len(df):,} ({100*matched.mean():.1f}%)")
print(f"unmatched: {(~matched).sum():,}")
print(f"distinct SchoolDistrict: {df['SchoolDistrict'].nunique()}")
print(df["SchoolDistrict"].value_counts().head(10))

matched:   108,237 of 142,750 (75.8%)
unmatched: 34,513
distinct SchoolDistrict: 363
SchoolDistrict
Los Angeles Unified          14193
NonUnified_San Diego          6584
NonUnified_Los Angeles        6179
San Diego Unified             3864
NonUnified_Orange             3450
NonUnified_San Bernardino     3432
Capistrano Unified            2637
Desert Sands Unified          2606
NonUnified_Riverside          2568
Palm Springs Unified          2322
Name: count, dtype: int64


In [28]:
# cross check with the raw dataset col of high school districts to validate the unified

raw = pd.read_csv(data[-1], usecols=["ListingKey","HighSchoolDistrict"])
chk = df.merge(raw, on="ListingKey", how="left")
print("HighSchoolDistrict present:", chk["HighSchoolDistrict"].notna().mean().round(3))
print("\nsample: spatial vs MLS district")
print(chk[["SchoolDistrict","HighSchoolDistrict"]].dropna().head(15).to_string())

HighSchoolDistrict present: 0.065

sample: spatial vs MLS district
                      SchoolDistrict               HighSchoolDistrict
7              Shandon Joint Unified        Paso Robles Joint Unified
21                   Oakland Unified                          Oakland
196   Big Oak Flat-Groveland Unified                   Gold Oak Union
261              Bear Valley Unified              Bear Valley Unified
305                  Needles Unified  Needles Unified School District
542                 NonUnified_Butte                    Oroville City
560             NonUnified_San Diego                  Escondido Union
569             NonUnified_San Diego                  Escondido Union
887             NonUnified_San Mateo                            Other
913                  Compton Unified                  Compton Unified
980                  Lynwood Unified                  Compton Unified
1053        Rim of the World Unified                 Rim of the World
1815                Non

## Bed/Bath Ratio

- Beds per bath in each property
==> BedroomsTotal / BathroomsTotalInteger
- all NaNs should've been dropped by this point of the run

In [29]:
df["BedBathRatio"] = df["BedroomsTotal"] / df["BathroomsTotalInteger"]

assert df["BedBathRatio"].isna().sum() == 0, "BedBathRatio has NaNs"
assert np.isfinite(df["BedBathRatio"]).all(), "BedBathRatio has inf (a zero-bath row slipped through)"
print(f"BedBathRatio: min {df['BedBathRatio'].min():.2f} | median {df['BedBathRatio'].median():.2f} | max {df['BedBathRatio'].max():.2f}")

BedBathRatio: min 0.14 | median 1.50 | max 7.50


### Interpret:

- The minimum ratio (0.14) is representative of a higher end home (more baths than bedrooms)
- Vice versa, the max value of (7.5) means more bedrooms than bathrooms

## Property Age
- looking into the age at the time of sale as reported in the data
- YearBuilt was already previously screened
- Calculated by Closing Date (date property sold to new owners) - Year Built

In [30]:
raw_age = df["CloseDate"].dt.year - df["YearBuilt"]
df["PropertyAge"] = (raw_age).clip(lower=0)

#check
assert df["PropertyAge"].isna().sum() == 0, "PropertyAge has NaNs"
assert (df["PropertyAge"] >= -2).all(), "potentially incorrect YearBuilt"
print(f"PropertyAge: min {df['PropertyAge'].min():.0f} | median {df['PropertyAge'].median():.0f} | max {df['PropertyAge'].max():.0f}")
print(f"negatives clipped to 0: {((df['CloseDate'].dt.year - df['YearBuilt']) < 0).sum()}")

PropertyAge: min 0 | median 49 | max 172
negatives clipped to 0: 14


### Interpret:
- Minimum age of 0 means that the home was built less than a year ago (new construction)

## LotSizeAcres to LivingArea Ratio

- investigating the land to actual livable area of the property

In [31]:
LS_sqft = df["LotSizeAcres"] * 43_560 # unit conversion 

df["LotToLiving"] = LS_sqft / df["LivingArea"]

assert df["LotToLiving"].isna().sum() == 0, "LotToLiving has NaNs"
assert np.isfinite(df["LotToLiving"]).all(), "LotToLiving has inf (zero LivingArea slipped through)"
print(f"LotToLiving: min {df['LotToLiving'].min():.2f} | median {df['LotToLiving'].median():.2f} | max {df['LotToLiving'].max():.2f}")
print(f"  p95: {df['LotToLiving'].quantile(0.95):.1f} | p99: {df['LotToLiving'].quantile(0.99):.1f}")

df["LotToLiving_log"] = np.log(df["LotToLiving"])
print(f"LotToLiving_log: min {df['LotToLiving_log'].min():.2f} | median {df['LotToLiving_log'].median():.2f} | max {df['LotToLiving_log'].max():.2f}")
print(f"  p95: {df['LotToLiving_log'].quantile(0.95):.2f} | p99: {df['LotToLiving_log'].quantile(0.99):.2f}")

LotToLiving: min 0.10 | median 4.22 | max 20808.98
  p95: 22.5 | p99: 134.8
LotToLiving_log: min -2.30 | median 1.44 | max 9.94
  p95: 3.11 | p99: 4.90


## LivingAreaPerBedroom

Comparing the liveable area to bedrooms
Calculating by: LivingArea / BedRoomsTotal

In [32]:
df["LivingAreaPerBedroom"] = df["LivingArea"] / df["BedroomsTotal"].clip(lower=1)

assert df["LivingAreaPerBedroom"].isna().sum() == 0, "LivingAreaPerBedroom has NaNs"
assert np.isfinite(df["LivingAreaPerBedroom"]).all(), "LivingAreaPerBedroom has inf"
print(f"LivingAreaPerBedroom: min {df['LivingAreaPerBedroom'].min():.0f} | "
      f"median {df['LivingAreaPerBedroom'].median():.0f} | max {df['LivingAreaPerBedroom'].max():.0f}")

LivingAreaPerBedroom: min 35 | median 542 | max 4287


## HasGarage

Capturing detached garages alongside the attached ones

In [33]:
df["HasGarage"] = (df["GarageSpaces"] > 0).astype(int)

assert df["HasGarage"].isin([0, 1]).all(), "HasGarage not binary"
print(f"HasGarage: {df['HasGarage'].mean()*100:.1f}% of rows have a garage")

HasGarage: 88.3% of rows have a garage


## Amenity Counts

Scoring the property by amount of amenities offered with the YN flags

In [34]:
AMENITY_YN = ["ViewYN", "PoolPrivateYN", "FireplaceYN", "BasementYN"]
df["AmenityCount"] = df[AMENITY_YN].fillna(False).astype(int).sum(axis=1)

assert df["AmenityCount"].between(0, len(AMENITY_YN)).all(), "AmenityCount out of range"
print(f"AmenityCount (of {len(AMENITY_YN)}): min {df['AmenityCount'].min()} | "
      f"median {df['AmenityCount'].median():.0f} | max {df['AmenityCount'].max()}")

AmenityCount (of 4): min 0 | median 1 | max 4


# Encoding

Converting the categorical fields to numeric

- Boolean Fields
- CountyOrParish

In [35]:
# Boolean Fields
bool_cols = [col for col in df.columns if col.endswith('YN')] + ['LotSizeAcres_imputed']
df[bool_cols] = df[bool_cols].astype(int)
print("Converted to 0/1 int:")
for col in bool_cols:
    print(f"  {col}")

Converted to 0/1 int:
  ViewYN
  BasementYN
  PoolPrivateYN
  AttachedGarageYN
  FireplaceYN
  NewConstructionYN
  LotSizeAcres_imputed


In [36]:
# CountyOrParish

null_county = df['CountyOrParish'].isna().sum()
if null_county:
    print(f'Filling {null_county} missing CountyOrParish values with "Unknown"')
    df['CountyOrParish'] = df['CountyOrParish'].fillna('Unknown')

##
fit_mask    = df['CloseDate'] < TEST_MONTH_START
county_cats = sorted(df.loc[fit_mask, 'CountyOrParish'].unique())
county_col  = df['CountyOrParish'].where(df['CountyOrParish'].isin(county_cats))
county_dummies = pd.get_dummies(county_col, prefix='County', dtype=int)
# county_dummies = pd.get_dummies(df['CountyOrParish'], prefix='County', dtype=int)
county_dummies.columns = [col.replace(' ', '_') for col in county_dummies.columns]
df = pd.concat([df, county_dummies], axis=1)

print(f'Encoded CountyOrParish dummy columns: {county_dummies.shape[1]}'
      f'(from {len(county_cats)} train categories)')

Encoded CountyOrParish dummy columns: 56(from 56 train categories)


In [37]:
# School District

fit_mask      = df['CloseDate'] < TEST_MONTH_START
district_cats = sorted(df.loc[fit_mask, 'SchoolDistrict'].unique())
n_test_only   = df.loc[~df['SchoolDistrict'].isin(district_cats), 'SchoolDistrict'].nunique()
district_col  = df['SchoolDistrict'].where(df['SchoolDistrict'].isin(district_cats))
district_dummies = pd.get_dummies(district_col, prefix='District', dtype=int)
district_dummies.columns = [c.replace(" ", "_") for c in district_dummies.columns]


df = pd.concat([df, district_dummies], axis=1)

row_sums = district_dummies.sum(axis=1)
assert set(row_sums.unique()) <= {0, 1}, "district dummies not one-hot"

print(f"District one-hot columns: {district_dummies.shape[1]}"
      f"(from {len(district_cats)} train categories; {n_test_only} test-only district(s) -> all-zero)")

print(f"rows with a known-train district: {(row_sums == 1).sum():,} | unknown: {(row_sums == 0).sum():,}")

District one-hot columns: 362(from 362 train categories; 1 test-only district(s) -> all-zero)
rows with a known-train district: 142,749 | unknown: 1


In [38]:
# Postal Code
# rare groupings
RARE_MIN = 50
fit_mask = df['CloseDate'] < TEST_MONTH_START
df['PostalCode'] = df['PostalCode'].fillna('Unknown').astype(str).str.strip()

zc = df.loc[fit_mask, 'PostalCode'].value_counts()
zip_cats = sorted(zc[zc >= RARE_MIN].index)
zip_col  = df['PostalCode'].where(df['PostalCode'].isin(zip_cats), 'Other')
zip_dummies = pd.get_dummies(zip_col, prefix='Zip', dtype=int)
zip_dummies.columns = [c.replace(' ', '_') for c in zip_dummies.columns]

df = pd.concat([df, zip_dummies], axis=1)

print(f"Zip one-hot columns: {zip_dummies.shape[1]} (from {len(zip_cats)} train ZIPs >= {RARE_MIN} sales; rest -> Zip_Other)")

Zip one-hot columns: 705 (from 704 train ZIPs >= 50 sales; rest -> Zip_Other)


In [39]:
# City

df['City'] = df['City'].fillna('Unknown').astype(str).str.strip().str.title()
cc = df.loc[fit_mask, 'City'].value_counts()
city_cats = sorted(cc[cc >= RARE_MIN].index)
city_col  = df['City'].where(df['City'].isin(city_cats), 'Other')
city_dummies = pd.get_dummies(city_col, prefix='City', dtype=int)
city_dummies.columns = [c.replace(' ', '_') for c in city_dummies.columns]

df = pd.concat([df, city_dummies], axis=1)


print(f"City one-hot columns: {city_dummies.shape[1]} (from {len(city_cats)} train cities >= {RARE_MIN} sales; rest -> City_Other)")

City one-hot columns: 407 (from 406 train cities >= 50 sales; rest -> City_Other)


Notes:

- i. Boolean Fields
    - Converted 7 Boolean variables (YN fields and LotSizeAcres_imputed) from Boolean values to 0/1 integer format.
- ii. CountyOrParish
    - One-hot encoded CountyOrParish into 56 binary indicator columns for use in modeling.
    - Standardized dummy column names by replacing spaces with underscores.
- iii. SchoolDistrict
    - mirrored CountyOrParish one-hot

## Normalizing

ClosePrice and LivingArea
LotSizeAcres
BedroomsTotal / BathroomsTotalInterval


In [40]:
results = []

for col in ['ClosePrice', 'LivingArea', 'LotSizeAcres']:
    raw_skew = df[col].skew()
    log1p_skew = np.log1p(df[col]).skew()
    log_skew   = np.log(df[col]).skew()
    print(f'{col:<15} raw: {raw_skew:8.2f} | log1p: {log1p_skew:8.2f} | log: {log_skew:8.2f}')

df['ClosePrice_log']   = np.log1p(df['ClosePrice'])
df['LivingArea_log']   = np.log1p(df['LivingArea'])
df['LotSizeAcres_log'] = np.log(df['LotSizeAcres'])

print("\nAdded normalized features:")
print("  - ClosePrice_log")
print("  - LivingArea_log")
print("  - LotSizeAcres_log")

ClosePrice      raw:    12.00 | log1p:     0.52 | log:     0.52
LivingArea      raw:     4.16 | log1p:     0.34 | log:     0.33
LotSizeAcres    raw:    60.26 | log1p:     5.57 | log:     2.19

Added normalized features:
  - ClosePrice_log
  - LivingArea_log
  - LotSizeAcres_log


Notes:
- Applied a log transformation to reduce right-skewed distributions
- Creatied the features ClosePrice_log, LivingArea_log, and LotSizeAcres_log.
- The transformation substantially reduced skew for:
    - ClosePrice (12.00 --> 0.52), 
    - LivingArea (4.16 --> 0.33)
- LotSizeAcres remained positively skewed (60.26 --> 2.19), indicating extreme outliers remain

## Train/Test Split
- Using the most recent month of available data as the test set, and the X months immediately preceding it as the training set. (not hardcoded)
- X is not fixed, treat the training window length as a tunable choice and experiment to determine the optimal value of X.

In [41]:
# TRAIN_WINDOW_MONTHS = 3   # <-- X: the tunable training-window length
TRAIN_WINDOW_MONTHS = min(11, df['CloseDate'].dt.to_period('M').nunique() - 1)

df['CloseMonth'] = df['CloseDate'].dt.to_period('M')
months = sorted(df['CloseMonth'].unique())
print(f'Months present: {[str(m) for m in months]}')

test_month = months[-1]
print(f'Test month: {test_month}')


def chrono_split(frame, x, test_month=test_month):
    """Most recent month = test set; the x months immediately preceding = train set."""
    train_months = [test_month - i for i in range(x, 0, -1)]
    train = frame[frame['CloseMonth'].isin(train_months)].copy()
    test = frame[frame['CloseMonth'] == test_month].copy()
    return train, test

Months present: ['2025-06', '2025-07', '2025-08', '2025-09', '2025-10', '2025-11', '2025-12', '2026-01', '2026-02', '2026-03', '2026-04', '2026-05', '2026-06']
Test month: 2026-06


In [42]:
# Demonstrate the trade-off across every possible X
for x in range(1, len(months)):
    tr, te = chrono_split(df, x)
    print(f'X={x}: train {tr["CloseMonth"].min()} .. {tr["CloseMonth"].max()} '
          f'({len(tr):,} rows)  |  test {test_month} ({len(te):,} rows)')

X=1: train 2026-05 .. 2026-05 (11,997 rows)  |  test 2026-06 (12,837 rows)
X=2: train 2026-04 .. 2026-05 (24,008 rows)  |  test 2026-06 (12,837 rows)
X=3: train 2026-03 .. 2026-05 (35,163 rows)  |  test 2026-06 (12,837 rows)
X=4: train 2026-02 .. 2026-05 (43,693 rows)  |  test 2026-06 (12,837 rows)
X=5: train 2026-01 .. 2026-05 (51,158 rows)  |  test 2026-06 (12,837 rows)
X=6: train 2025-12 .. 2026-05 (61,593 rows)  |  test 2026-06 (12,837 rows)
X=7: train 2025-11 .. 2026-05 (71,304 rows)  |  test 2026-06 (12,837 rows)
X=8: train 2025-10 .. 2026-05 (83,304 rows)  |  test 2026-06 (12,837 rows)
X=9: train 2025-09 .. 2026-05 (94,732 rows)  |  test 2026-06 (12,837 rows)
X=10: train 2025-08 .. 2026-05 (106,155 rows)  |  test 2026-06 (12,837 rows)
X=11: train 2025-07 .. 2026-05 (118,247 rows)  |  test 2026-06 (12,837 rows)
X=12: train 2025-06 .. 2026-05 (129,913 rows)  |  test 2026-06 (12,837 rows)


In [43]:
train_df, test_df = chrono_split(df, TRAIN_WINDOW_MONTHS)
print(f'Chosen X = {TRAIN_WINDOW_MONTHS}')
print(f'Train: {train_df.shape[0]:,} rows   Test: {test_df.shape[0]:,} rows')

# Chronological sanity check: no train row may close on/after the test month
assert train_df['CloseMonth'].max() < test_month
assert (test_df['CloseMonth'] == test_month).all()

Chosen X = 11
Train: 118,247 rows   Test: 12,837 rows


Note:
- Used the most recent month as test set and evaluated training windows preceding the test period.
- Training window is tunable
- Checked integrity by ensuring training observations occurred before the test month

## Export Cleaned CSV

- Not visible in the public Github repo for privacy. 
- Exported to local machine.

In [44]:
start_month = df['CloseDate'].min().strftime('%Y%m')
end_month   = df['CloseDate'].max().strftime('%Y%m')

OUTPUT_PATH = f"{Data_dir}/cleaned_CRMLSSold{start_month}_{end_month}.csv"

df.to_csv(OUTPUT_PATH, index=False)

print(f'Wrote {OUTPUT_PATH}')
print(f'Final shape: {df.shape[0]:,} rows x {df.shape[1]} columns')
print(f'\nColumns:')
for col in df.columns:
    print(f'  {col}')

Wrote ../../data/cleaned_CRMLSSold202506_202606.csv
Final shape: 142,750 rows x 1567 columns

Columns:
  ListingKey
  CloseDate
  ClosePrice
  LivingArea
  BedroomsTotal
  BathroomsTotalInteger
  LotSizeAcres
  LotSizeAcres_imputed
  YearBuilt
  Stories
  GarageSpaces
  ParkingTotal
  ViewYN
  BasementYN
  PoolPrivateYN
  AttachedGarageYN
  FireplaceYN
  NewConstructionYN
  CountyOrParish
  City
  PostalCode
  ClosePrice_repaired
  Latitude
  Longitude
  YearBuilt_imputed
  SchoolDistrict
  BedBathRatio
  PropertyAge
  LotToLiving
  LotToLiving_log
  LivingAreaPerBedroom
  HasGarage
  AmenityCount
  County_Alameda
  County_Amador
  County_Butte
  County_Calaveras
  County_Colusa
  County_Contra_Costa
  County_El_Dorado
  County_Fresno
  County_Glenn
  County_Humboldt
  County_Imperial
  County_Inyo
  County_Kern
  County_Kings
  County_Lake
  County_Lassen
  County_Los_Angeles
  County_Madera
  County_Marin
  County_Mariposa
  County_Mendocino
  County_Merced
  County_Modoc
  County_Mo

- Cleaned CSV (`cleaned_CRMLSSold.csv`) is written into the gitignored `/data` folder, so the cleaned data also stays off the public repo
- The final dataset contains scales according to months loaded, including repaired, encoded, and transformed features.

## Findings and Final Statements

This file will be applicable to any CRMLSSold*.csv file for cleaning.
Final .csv is NOT uploaded to the github.

### Summary
- Successfully developed a reusable preprocessing pipeline for CRMLSSold datasets.
- The pipeline performs filtering, data quality checks, missing value handling, feature selection, encoding, normalization, and chronological train/test splitting.
- Cleaning decisions are data-driven where possible (e.g., dynamic missing-value detection, Boolean detection, leakage prevention, and county encoding), allowing the notebook to generalize across future `CRMLSSold*.csv` files.

### Output
- Row and column counts scale with the months loaded. The current full-window extract (Jun 2025 – Jun 2026) dataframe: about 142,750 rows × 1,567 columns.
- Exported to: `../data`
- Dataset includes repaired, encoded, and transformed features while preserving audit columns for traceability in upcoming tasks

### Limitations
- This pipeline assumes the input follows the CRMLSSold schema.
- Thresholds (e.g., unit repair ratios, outlier bounds, training window size) are configurable and may require adjustment for future datasets or modeling objectives.
- Naming of the export file assumes the files used are consecutive months.